topology/probes/ovs_of.go

```diff
@@ -0,0 +1,503 @@
+		return rule, errors.New("Rule syntax split filter and actions")
+	}
+	return rule, nil
+}
+
+func makeFilter(rule *Rule) string {
+	if rule.Filter == "" {
+		return fmt.Sprintf("table=%d", rule.Table)
+	}
+	return fmt.Sprintf("table=%d,%s", rule.Table, rule.Filter)
+}
+
+// Execute exposes an interface to command launch on the OS
+type Execute interface {
+	ExecCommand(string, ...string) ([]byte, error)
+	ExecCommandPipe(context.Context, string, ...string) (io.Reader, error)
+}
+
+// RealExecute is the actual implementation given below. It can be overriden for tests.
+type RealExecute struct{}
+
+var executor Execute = RealExecute{}
+
+// ExecCommand executes a command on a host
+func (r RealExecute) ExecCommand(com string, args ...string) ([]byte, error) {
+	var command *exec.Cmd
+	/* #nosec */
+	command = exec.Command(com, args...)
+	bytes, err := command.CombinedOutput()
+	return bytes, err
+}
+
+// ExecCommandPipe executes a command on a host and gives back a pipe to control it.
+func (r RealExecute) ExecCommandPipe(ctx context.Context, com string, args ...string) (io.Reader, error) {
+	/* #nosec */
+	command := exec.CommandContext(ctx, com, args...)
+	out, err := command.StdoutPipe()
+	command.Stderr = command.Stdout
+	if err != nil {
+		return out, err
+	}
+	err = command.Start()
+	return out, err
+}
+
+// LaunchOnSwitch launches a command on a given switch
+func LaunchOnSwitch(cmd []string) (string, error) {
+	bytes, err := executor.ExecCommand(cmd[0], cmd[1:]...)
+	if err == nil {
+		return string(bytes), nil
+	}
+	return "", err
+}
+
+// LaunchContinuousOnSwitch launches  a stream producing command on a given switch
+func LaunchContinuousOnSwitch(ctx context.Context, cmd []string) (<-chan string, error) {
+	var cout = make(chan string, 10)
+	out, err := executor.ExecCommandPipe(ctx, cmd[0], cmd[1:]...)
+	if err != nil {
+		close(cout)
+		return cout, err
+	}
+	reader := bufio.NewReader(out)
+	var line string
+
+	go func() {
+		for {
+			line, err = reader.ReadString('\n')
+			if err == io.EOF {
+				close(cout)
+				break
+			} else if err != nil {
+				logging.GetLogger().Errorf("IO Error on command %s: %s", cmd[0], err.Error())
+			} else {
+				cout <- line
+			}
+		}
+	}()
+
+	return cout, nil
+}
+
+func countElements(filter string) int {
+	if len(filter) == 0 {
+		return 0
+	}
+	return len(strings.Split(filter, ","))
+}
+
+// completeRule completes the rule by looking at it again but with dump-flows. This gives back more elements such as priority.
+func completeRule(o *OvsOfProbe, event *Event) error {
+	oldrule := event.Rule
+	bridge := event.Bridge
+	// We want exactly n+1 items where n was the number of items in old filters. The reason is that now
+	// the priority is provided. Another approach would be to use the shortest filter as it is the more generic
+	expected := countElements(oldrule.Filter) + 1
+	filter := makeFilter(oldrule)
+	command, err1 := o.makeCommand([]string{"ovs-ofctl", "dump-flows"}, bridge, filter)
```

Comment:
From the point of vue of the extension and ovs-ofctl, an action or a filter is just a string separated by commas and is arlready ready to be given to the end-user. Only a few of them are parsed for giving better feedback (port filter, output on a port, drop, resubmit...) . At the level of Openflow these are just random sequence of bytes in a TLV format that MUST be parsed to be meaningful to the end-user. So we are much more dependent on changes at the protocol level than at the text level.